designs/2024-config-extends/README.md

```diff
@@ -0,0 +1,613 @@
+
+The [`normalize()`](https://github.com/eslint/eslint/blob/fd33f1315ac59b1b3828dbab8e1e056a1585eff0/lib/config/flat-config-array.js#L141) and [`normalizeSync()`](https://github.com/eslint/eslint/blob/fd33f1315ac59b1b3828dbab8e1e056a1585eff0/lib/config/flat-config-array.js#L160) methods will be updated to use a `ConfigExtender` instance to evaluate any `extends` keys. To do this, we'll need to evaluate each object for `extends` before allowing calling the superclass method. Here's an example of what `normalize()` will look like:
+
+```js
+
+const extender = new ConfigExtender();
+
+class FlatConfigArray {
+
+    // snip
+
+    normalize(context) {
+
+        // make sure not to make any changes if already normalized
+        if (this.isNormalized()) {
+            throw new Error("...");
+        }
+
+        // replace each element with an array
+        this.forEach((element, i) => {
+            const configs = Array.isArray(element) ? element : [element];
+            this[i] = configs.flat().map(config => extender.evaluate(config))
+        });
+
+        // proceed as usual
+        return super.normalize(context)
+            .catch(error => {
+                if (error.name === "ConfigError") {
+                    throw wrapConfigErrorWithDetails(error, this[originalLength], this[baseLength]);
+                }
+
+                throw error;
+
+            });
+    }
+}
+```
+
+## Documentation
+
+This will require documentation changes and an introductory blog post.
+
+At a minimum, these pages will have to be updated:
+
+* https://eslint.org/docs/latest/use/getting-started#configuration
+* https://eslint.org/docs/latest/use/configure/configuration-files
+* https://eslint.org/docs/latest/extend/plugins#configs-in-plugins
+* https://eslint.org/docs/latest/use/configure/combine-configs
+
+## Drawbacks
+
+1. Introducing a new key to flat config at this point means that users of ESLint v8 won't be able to use it. Even though v8 is EOL, we had made an attempt to provide forward compatibility to make people's transition easier.
+1. While this may reduce complexity for users, it increases the complexity of config evaluation inside of ESLint. This necessarily means a performance cost that we won't be able to quantify until implementation.
+1. Users will have to know which version of ESLint supports `extends` in order to use it. There really isn't an easy way to feature test this capability other than to inspect the version of the current ESLint package.
+
+## Backwards Compatibility Analysis
+
+This proposal is additive and does not affect the way existing configurations are evaluated.
+
+## Alternatives
+
+
+### A `defineConfig()` function
+
+One alternative is to create a `defineConfig()` function that is exported from the `eslint` package. All of the functionality described in this RFC could be implemented in that function, leaving both `ConfigArray` and `FlatConfigArray` in their current state without the need for changes.
```

Comment:
> that is exported from the `eslint` package

One suggestion -- export it from a new package like `@eslint/define-config` which should be a dependency of `eslint`, and `eslint` should re-export it.

The reason I suggest this is that it means that if the plan is to add logic into `defineConfig` to allow utilities that make configs easier to use, then allowing a user to install and upgrade `@eslint/define-config` independently of the `eslint` version would mean you can easily "backport" config changes without forcing the user to upgrade their ESLint version.

I've seen a number of users that are locked on a specific ESLint version for various reasons (sometimes political/organisational problems that block them, and sometimes it's just hard to upgrade as a new version introduces new errors that will take time to fix). Being able to just install/bump `@eslint/define-config` would mean they can get config improvements without having to fix new reports and the like.

Feasibly it would also be possible for the latest version of `@eslint/define-config` to be backwards compatible across major releases -- meaning people could get the benefit of `extends` and other improvements on a previous major release. Major upgrades are often a real hurdle esp at large organisations.

This has been one of the benefits of `tseslint.config` -- it has existed outside of eslint's version so users have been able to use it for any ESLint version and get its benefits regardless of their underlying version -- including if they wanted to try flat configs on our minimum eslint version -- `8.57.0`.

Reply:
Yup, this is covered under the "Benefits" section below. I'll consider this a vote in favor of that option. :+1: